frontend/src/main/scala/bloop/engine/BuildLoader.scala

```diff
@@ -22,56 +27,136 @@ object BuildLoader {
+      inMemoryChanged: List[Build.InvalidatedInMemoryProject],
+      settingsForLoad: Option[WorkspaceSettings],
       logger: Logger
-  ): Task[List[Project]] = {
-    logger.debug(s"Loading ${configFiles.length} projects from '${configDir.syntax}'...")(
-      DebugFilter.Compilation
-    )
+  ): Task[List[LoadedProject]] = {
+    val incrementalLoadTask = Task {
+      val loadMsg = s"Loading ${configs.length} projects from '${configDir.syntax}'..."
+      logger.debug(loadMsg)(DebugFilter.All)
+      val rawProjects = configs.map(f => Task(loadProject(f.bytes, f.origin, logger)))
+      val groupTasks = rawProjects.grouped(10).map(group => Task.gatherUnordered(group)).toList
+      val newOrModifiedRawProjects = Task.sequence(groupTasks).map(fp => fp.flatten)
+
+      newOrModifiedRawProjects.flatMap { projects =>
+        val projectsRequiringMetalsTransformation = projects ++ inMemoryChanged.collect {
+          case Build.InvalidatedInMemoryProject(project, changes)
+              if changes.contains(WorkspaceSettings.SemanticDBVersionChange) =>
+            project
+        }
+
+        settingsForLoad match {
+          case None =>
+            Task.now(projectsRequiringMetalsTransformation.map(LoadedProject.RawProject(_)))
+          case Some(settings) =>
+            resolveSemanticDBForProjects(
+              projectsRequiringMetalsTransformation,
+              configDir,
+              settings.semanticDBVersion,
+              settings.supportedScalaVersions,
+              logger
+            ).map { transformedProjects =>
+              transformedProjects.map {
+                case (project, None) => LoadedProject.RawProject(project)
+                case (project, Some(original)) =>
+                  LoadedProject.ConfiguredProject(project, original, settings)
+              }
+            }
+        }
+      }
+    }
 
-    val all = configFiles.map(f => Task(Project.fromBytesAndOrigin(f.bytes, f.origin, logger)))
-    val groupTasks = all.grouped(10).map(group => Task.gatherUnordered(group)).toList
-    Task.sequence(groupTasks).map(_.flatten).executeOn(ExecutionContext.ioScheduler)
+    incrementalLoadTask.flatten.executeOn(ExecutionContext.ioScheduler)
   }
 
   /**
-   * Load all the projects from `configDir` in a parallel, lazy fashion via monix Task.
+   * Resolves the semanticdb plugin for every project with a different Scala
+   * version and then enables the Metals settings on those projects that have
+   * changed in this incremental build load.
    *
-   * @param configDir The base directory from which to load the projects.
-   * @param logger The logger that collects messages about project loading.
-   * @return The list of loaded projects.
+   * Some of the logic of this method has a duplicate version in
+   * `loadSynchronously` that doesn't use `Task` to resolve SemanticDB plugins
+   * in parallel and apply the Metals projects. Any change here **must** also
+   * be propagated there.
    */
-  def load(
```

Comment:
The ordinary load that loads everything isn't implemented in terms of this but in terms of `checkForChange`, so that's why I removed this method.